diff --git a/core/lib/Drupal/Core/StringTranslation/PluralTranslatableMarkup.php b/core/lib/Drupal/Core/StringTranslation/PluralTranslatableMarkup.php
index 4600c17..6363d6b 100644
--- a/core/lib/Drupal/Core/StringTranslation/PluralTranslatableMarkup.php
+++ b/core/lib/Drupal/Core/StringTranslation/PluralTranslatableMarkup.php
@@ -132,7 +132,13 @@ public function render() {
         // If the index cannot be computed or there's no translation, use the
         // second plural form as a fallback (which allows for most flexibility
         // with the replaceable @count value).
-        $return = $translated_array[1];
+        if (isset($translated_array[1])) {
+          $return = $translated_array[1];
+        }
+        else {
+          $translated_array = explode(static::DELIMITER, $this->string);
+          $return = isset($translated_array[1]) ? $translated_array[1] : $translated_array[0];
+        }
       }
     }
 
